Feature/221002 s3 setting

beetrbgus opened the pull request:
S3 파일 업로드 기능 추가 및 설정 추가
application-aws.yml 파일의  accessKey와 secretKey는 Notion에 공유 됨

beetrbgus committed:
S3 파일 업로드 설정

- local, prod 환경에서 aws 프로퍼티 프로필 등록
- s3에 bucket 연결
- gradle.properties 버전추가

beetrbgus committed:
S3 파일 업로드 기능

- S3 Java 설정
- S3에 파일 업로드 기능 및 Controller 생성

beetrbgus committed:
S3 파일 정보로 Skill Entity 생성

- Entity 제작
- Skill 등록 Controller, Service

yunhozz commented on src/main/java/com/justcodeit/moyeo/study/persistence/Skill.java:
```diff
@@ -0,0 +1,55 @@
+
+@Entity
+@Table(name = "skill")
+@RequiredArgsConstructor
```
생성자 주입을 하는 곳이 없어 보이는데 이 어노테이션을 사용한 이유가 있을까요?

beetrbgus replied:
무슨 말이죵???

yunhozz replied:
@RequiredArgsConstructor 어노테이션은 보통 외부에서 의존성을 주입해주기 위해 사용된다고 알고 있는데 final이나 @NotNull이 붙은 필드가 보이지 않아서요

mycode01 replied:
@yunhozz `@RequiredArgsContstructor`는 그냥 final이나 `@NonNull`필드에 대한 생성자를 만들어주는 어노테이션입니다. 저도 방금 검색해봤는데 언제부터 'spring에서 생성자 주입을 하기 위한 어노테이션'이라고 와전됐는지 모르겠네요.
스프링이나 di랑은 전혀 상관없습니다.

mycode01 replied:
위 지적과는 별개로 `@RequiredArgsConstructor`가 쓰일 이유가 없는건 맞는거 같습니다.
어노테이션을 제거한다면 private 기본생성자를 추가바랍니다.

yunhozz replied:
@mycode01 좋은 의견 감사드립니다. 어떤 의미인지는 인지하고 있었지만 스프링에 국한해서 생각했었네요..

beetrbgus replied:
@mycode01 궁금한게 JPA에서 객체 조회할 때 Proxy로 감싸서 오기 때문에 기본 생성자는 protected 또는 public이 있어야 된다고 알고 있는데 private으로 사용 가능한가요??

mycode01 replied:
jpa는 내부적으로 reflection 을 이용하여 객체를 생성합니다. reflection은 private 메소드도 접근에 가능하기 때문에 private 가 이 경우에는 적절할거같습니다.
물론 protected 도 괜찮습니다.

beetrbgus replied:
추후 확인 요망 ( private 기본 생성자 만들면 실행은 되는데 Intellij에서는 Error 문구 뜸 )

yunhozz commented:
통일성을 위해 indent를 2로 맞춰주시면 감사하겠습니다!

jjoylee commented on src/main/java/com/justcodeit/moyeo/study/interfaces/resource/S3Controller.java:
```diff
@@ -0,0 +1,20 @@
+
+    @PostMapping
+    public void uploadSkill() {
+        String path = "C:\\Users\\beetr\\OneDrive\\바탕 화면\\skill stack\\svg";
```
path는 임시인 것인가요? 개인 컴퓨터 Url인 것 같아서 이렇게 설정해도 되는지 궁금합니다. 그리고 upload2는 이름을 명확히 하는 것이 좋을 것 같습니다.

beetrbgus replied:
이름체크 확인용 근데 한번 쓰고 말거라서 일단  저렇게 적어놓고 지금은 파라미터로 String path를 받을 수 있게 수정했습니다.

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/interfaces/dto/s3/SkillService.java:
```diff
@@ -0,0 +1,47 @@
+        }
+        return response;
+    }
+}
```
클래스의 패키지가 부적절한거 같습니다.

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/interfaces/dto/s3/SkillService.java:
```diff
@@ -0,0 +1,47 @@
+    private final SkillRepository skillRepository;
+    private final S3Service s3Service;
+
+    public BaseResponse skillSave() {
```
BaseResponse가 interface dto 계층에 있는 이유는 사용자와 컨트롤러 사이의 레이어에서 사용하기 위해서입니다. Service <> Controller <> 사용자 를 가로지르기 위함이 아닙니다.

beetrbgus replied:
확인요 BaseResponse는 컨트롤러 단에서 넣는게 맞겠네요

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/interfaces/dto/s3/SkillService.java:
```diff
@@ -0,0 +1,47 @@
+
+    public BaseResponse skillSave() {
+        BaseResponse response;
+        try {
```
적절한 추상화 수준을 유지한 코드를 작성하기 바랍니다.
무슨 말인지 잘 모르겠다면 다음을 참고 바랍니다.
https://dev-splin.github.io/clean%20code/CleanCode-Chapter3/#3-3-%ED%95%A8%EC%88%98-%EB%8B%B9-%EC%B6%94%EC%83%81%ED%99%94-%EC%88%98%EC%A4%80%EC%9D%80-%ED%95%98%EB%82%98%EB%A1%9C

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/interfaces/resource/S3Controller.java:
```diff
@@ -0,0 +1,20 @@
+        String path = "C:\\Users\\beetr\\OneDrive\\바탕 화면\\skill stack\\svg";
+        s3Service.upload2(path);
+    }
+}
```
이 클래스가 해당 PR에 존재할 이유를 잘 모르겠습니다.
단발성 업로드를 위한 부분은 제외 바랍니다.

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/model/aws/s3/S3Config.java:
```diff
@@ -0,0 +1,39 @@
+                .withS3Client(amazonS3Client())
+                .build();
+    }
+}
```
클래스의 패키지가 부적절한거 같습니다.

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/model/aws/s3/S3Service.java:
```diff
@@ -0,0 +1,71 @@
+
+        return s3FileLocationList;
+    }
+}
```
클래스의 패키지가 부적절한거 같습니다.

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/model/aws/s3/S3Service.java:
```diff
@@ -0,0 +1,71 @@
+     * S3에 로컬 폴더를 업로드 하는 코드.
+     * S3의 폴더 구조를 로컬 C 드라이브의 폴더에 맞추기 위해서 virtualDirectoryKeyPrefix를 ""로 설정
+     */
+    public void upload2(String path) {
```
단발성 업로드를 위한 코드는 제외 바랍니다.

beetrbgus replied:
확인요-

mycode01 commented on src/main/resources/application-aws.yml:
```diff
@@ -0,0 +1,13 @@
+  aws:
+    credentials:
+      accessKey: # S3 accessKey
+      secretKey: # S3 secretKey
```
파일을 한번 git에 추가하고나면 특별한 방법을 사용하지 않고서는 계속 tracking 됩니다.
무슨말이냐면 다른 구성원들이 accessKey에 notion에서 올려둔 키를 보고 수정을 하게 되면 git은 해당파일이 변경된걸로 판단하고 다음번 커밋때 자동으로 해당 파일을 커밋 대상 파일로 인식하게 됩니다.
그럼 accessKey의 퍼블릭 노출을 피하기 위해 커밋 할때마다 이 파일을 커밋 대상에서 제외해야하고, 
실수로 누군가 커밋대상에서 제외하지 않고 커밋 후 푸시를 하게 되면 accessKey는 해당 리비전에 계속해서 남게 됩니다. 커밋된 리비전은 undo가 불가능합니다.
해당 파일을 그냥 notion에 공유하는게 나을거같습니다.

mycode01 commented on build.gradle:
```diff
@@ -40,6 +42,9 @@ dependencies {
 	testAnnotationProcessor 'org.projectlombok:lombok'
+
+	implementation 'com.amazonaws:aws-java-sdk-s3:1.12.311'
+	implementation 'org.apache.commons:commons-text:1.9'
```
사용되지 않는 라이브러리 의존은 제거바랍니다.

mycode01 commented on build.gradle:
```diff
@@ -25,6 +25,8 @@ dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-security'
 	implementation 'org.springframework.boot:spring-boot-starter-web'
 	implementation 'org.springframework.boot:spring-boot-starter-thymeleaf' // 임시 oauth 테스트용
+	compile 'org.springframework.cloud:spring-cloud-starter-aws:2.2.6.RELEASE'
```
사용되지 않는 라이브러리 의존은 제거바랍니다.
제 경험상 특히 spring-cloud 쪽 프레임워크는 다른 라이브러리랑 충돌이 잦습니다.

beetrbgus replied:
28라인은 s3 때문에 계속 사용할 것 같은데 thymeleaf 말인가요??

mycode01 replied:
28라인 필요 없습니다.

beetrbgus committed:
PR Comment 수정

- 1회성 메소드 삭제
- S3 패키징 수정

beetrbgus committed:
PR Comment 수정

- Skill Required 생성자 변경 (Private 생성자는 JPA에서 Proxy 객체로 가져와야 하기 때문에 불가)

- 추상화 정도 높이기 위해 컨버터 생성

beetrbgus committed:
PR Comment 수정

- s3 yml 파일 제거 ( Notion에 공유 예정)

beetrbgus committed:
PR Comment 수정

- cloud start 패키지 의존성 제거
- lombok, mapstruct 버전 명시

jjoylee commented on src/main/java/com/justcodeit/moyeo/study/application/aws/s3/S3Service.java:
```diff
@@ -0,0 +1,72 @@
+            s3FileLocationList.add(s3FileLocation);
+        }
+        return s3FileLocationList;
+    }
```
`return amazonS3.listObjectsV2(bucket).getObjectSummaries().stream()
                .map(s3Summary -> s3Summary.getKey())
                .collect(Collectors.toList());`
이런식으로 람다 사용할 수 도 있을 것 같습니다.

jjoylee replied:
`
return amazonS3.listObjectsV2(bucket).getObjectSummaries().stream()
                .map(s3Summary -> s3Summary.getKey())
                .collect(Collectors.toList());
`
람다식으로 간단하게 표현할 수 있을 것 같습니다.

jjoylee commented on src/main/java/com/justcodeit/moyeo/study/application/skill/SkillService.java:
```diff
@@ -0,0 +1,39 @@
+                String name = fileFolderAndName[1].replaceAll(".svg","");
+
+                SkillCategory skillCategory = SkillCategoryConverter.getForFolderName(folder);
+                String url = "" + skillCategory.getEngWord();
```
"" + 는 없어도 괜찮을 것 같습니다.

beetrbgus replied:
AWS S3 Endpoint 를 넣었어야 ㅎ했는데 누락

yunhozz commented on src/main/java/com/justcodeit/moyeo/study/application/aws/s3/S3Service.java:
```diff
@@ -26,14 +26,17 @@ public class S3Service {
     private final AmazonS3 amazonS3;
     private final TransferManager transferManager;
 
+    @Value("${s3EndPoint}")
+    private final String S3_ENDPOINT;
```
final 을 써도 되는지 확인 부탁드려요~
